### PR TITLE
Enriched element

### PR DESCRIPTION
**Pull request by miklos1**
The FInAT version of `FIAT.EnrichedElement`. Related to #19.

Requires firedrakeproject/tsfc#131.

### CODE REVIEW

File: finat/enriched.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        results = [element.point_evaluation(order, refcoords, entity)
+                   for element in self.elements]
+        return self._compose_evaluations(results)
+
+
+def tree_map(f, *args):
+    """Like the built-in :py:func:`map`, but applies to a tuple tree."""
+    nonleaf, = set(isinstance(arg, tuple) for arg in args)
+    if nonleaf:
+        ndim, = set(map(len, args))
```
Comment (wence-) [inline finat/enriched.py]: assigned to but never used?

File: finat/enriched.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        def merge(tables):
+            tables = tuple(tables)
+            zeta = self.get_value_indices()
+            tensors = []
+            for elem, table in zip(self.elements, tables):
+                beta_i = elem.get_indices()
+                tensors.append(gem.ComponentTensor(
+                    gem.Indexed(table, beta_i + zeta),
+                    beta_i
+                ))
```
Comment (wence-) [inline finat/enriched.py]: So each sub element provides a `CT(table[beta_i, zeta], beta_i)`, which has free indices zeta (the indices of the outer element)

File: finat/enriched.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            beta = self.get_indices()
+            return gem.ComponentTensor(
+                gem.Indexed(gem.Concatenate(*tensors), beta),
+                beta + zeta
+            )
```
Comment (wence-) [inline finat/enriched.py]: And then you concatenate the sub element tables together and bind to the outer indices.

File: finat/enriched.py (hunk 1 above)
Comment (miklos1) [inline finat/enriched.py]: I only want the side effect actually, that is, a failure when arguments have different length. Perhaps
```python
assert len(set(map(len, args))) == 1
```
communicates the intention better, though I find the `len(set(map(len` part a bit inconvenient to read.

Comment (wence-) [inline finat/enriched.py]: I see, the latter is clearer I think (although I agree with the readability), and less likely to be removed at a later date.

**Pull request merged**